enzy_htp/analysis/cavity.py

```diff
@@ -0,0 +1,156 @@
+            * int: max_overlap value.
+    """
+    if sum(x is not None for x in (composing_residues, contain_ligand, target_cavity)) != 1:
+        raise ValueError("The `composing_residues`, `contain_ligand` and `target_cavity` are mutually exclusive to each other.")
+    
+    if (contain_ligand):
+        stru = cavity_list[0].stru
+        ligand_selection = select_stru(stru=stru, pattern=contain_ligand)
+        focus_point = np.array(ligand_selection.atoms[0].coord)
+        for cavity in cavity_list:
+            if cavity.contains(focus_point):
+                return cavity, 1
+            else:
+                continue
+        return None, 0
+    else:
+        focus_residue_keys = set()
+        cavity_similarity = dict()
+        if (target_cavity):
+            focus_residue_keys = set(resi.key() for resi in (target_cavity.boundary_residues + target_cavity.inner_residues))
+        if (composing_residues):
+            focus_residue_keys = set(resi.key() for resi in composing_residues)
+        for i, cavity in enumerate(cavity_list):
+            cavity_residue_keys = set(resi.key() for resi in (cavity.boundary_residues + cavity.inner_residues))
+            cavity_similarity[i] = len(cavity_residue_keys.intersection(focus_residue_keys))
+        return cavity_list[max(cavity_similarity, key=cavity_similarity.get)], max(cavity_similarity.values())
+
+def ensemble_cavity_volumes(
+        stru_esm: StructureEnsemble, 
+        composing_residues: List[Residue] = None, 
+        contain_ligand: str = None, 
+        target_cavity: Cavity = None, 
+        frame_0_based: bool = True,
+        **kwargs) -> List[float]:
+    """Caculate cavity volumes from a StructureEnsemble instance.
+    The cavity should be selected by a list of Residue instances, a ligand string (PyMOL formatted), or a Cavity instance.
+    
+    Args:
+        stru_esm (StructureEnsemble): The StructureEnsemble instance to calculate cavities from.
+        composing_residues (List[Residue], optional): A list of Residue instances composing the cavity. Mutually exclusive with `contain_ligand` or `target_cavity`.
+        contain_ligand (str, optional): The PyMOL-formatted pattern selecting the ligand in the cavity. Mutually exclusive with `target_cavity` or `composing_residues`.
+        target_cavity (Cavity, optional): The target cavity to track throughout the ensemble (by most overlapped residues). Mutually exclusive with `composing_residues` or `contain_ligand`.
+        frame_0_based (bool, optional): Indicate if the cavity selection input is formulated with the frame 0 structure to get target cavity.
+
+    Returns:
+        volumes (List[float]): The list of cavity volume value from the ensemble.
+    """
+    if sum(x is not None for x in (composing_residues, contain_ligand, target_cavity)) != 1:
+        raise ValueError("The `composing_residues`, `contain_ligand` and `target_cavity` are mutually exclusive to each other.")
```

Comment:
Put this message in _LOGGER and if the user didn't specify any one them this message reads confusing.